envoyproxy/envoy
# quic: refactor QUICHE flag implementation

QUICHE QUIC_FLAG implementation should be backed by Envoy runtime features not only because it's the canonical way (*1) but also for performance reasons.

It is observed that some quiche::TypedFlag access on the QUICHE datapath becomes a major scaling bottleneck when downstream QUIC traffic is heavy.

Simply replacing absl::MutexLock in TypedFlag implementation with absl::(Reader|Writer)MutexLock where possible is not satisfactory. The bottleneck would just move from the contention on the MutexLock to the  atomic operation on the refcount in the new ReaderMutexLock done by each Envoy worker thread. We need a truly scalable flag accessing. Unfortunately, Envoy::Runtime::runtimeFeatureEnabled is also not contention-free since the snapshot created by createNewSnapshot() is shared by every worker thread, which again leads to the refcount contention.

~~With this patch~~ **With this PR and #18601**, Envoy QUIC performance scales quite well. Please see the attached plot of the scalability comparison between Envoy (as a) QUIC server with-and-without this patch.

(*1) https://github.com/envoyproxy/envoy/blob/1b9c688997f5/source/common/quic/platform/quiche_flags_impl.h#L22-L24

<img src="https://user-images.githubusercontent.com/17020294/136225701-70be0d86-57cd-4335-8af1-a674263d6a6c.png" width="400">



---
(As a side note, added on 2021/10/11)

> Envoy::Runtime::runtimeFeatureEnabled is also not contention-free since the snapshot created by createNewSnapshot() is shared by every worker thread, which again leads to the refcount contention.

This is the reason why do we not only make QUICHE flag implementation backed by Envoy runtime features, but also "unshare" SnapshotImpl. But the contention really occurs? how to reproduce it? The way to reproduce it and its result are shown below:

* How to reproduce it?
  * build envoy server for two commits
    * https://github.com/envoyproxy/envoy/pull/18485/commits/56152d2070ac0d505e888a36cae33101f31c048b and https://github.com/envoyproxy/envoy/pull/18485/commits/d4f01ac927faad35a9768e6d75fcaa00471408bf
    * The latter just "unshares" SnapshotImpl, on top of the former commit.
    * The reason I use https://github.com/envoyproxy/envoy/pull/18485/commits/d4f01ac927faad35a9768e6d75fcaa00471408bf instead of https://github.com/envoyproxy/envoy/commit/87dffe0d1c9ec1695d5bc9147b8be16cf0c5ee0b is that it's a tiny diff and helps clarify the point. Rest assured that https://github.com/envoyproxy/envoy/commit/87dffe0d1c9ec1695d5bc9147b8be16cf0c5ee0b behaves well at the same level as https://github.com/envoyproxy/envoy/pull/18485/commits/d4f01ac927faad35a9768e6d75fcaa00471408bf.
    * Build `bazel.debug.server_only` with -O2 optimization (my ad-hoc diff: https://gist.github.com/lkpdn/cd416dbcb1a2dba061389bb8eff07f5c)
  * environment
    * launch two AWS EC2 c6gn.16xlarge instances (1 client + 1 server)
  * prepare a config file
    * see https://gist.github.com/lkpdn/170fac3ab5bbea3409ca79e58123f697
  * start an Envoy server with the config above
  * pin each worker thread onto a cpu core
  * stress the server from the client, using h2load (https://github.com/nghttp2/nghttp2/tree/quic)
    ```
    $ sudo docker run --rm -t --network=host localhost/h2load-quic -c 64 -t 64 -m 800 --warm-up-time=40s -D 40s --tls13-ciphers=TLS_AES_128_GCM_SHA256 --npn-list h3 --connect-to {{ server_ip }}:443 https://127.0.0.1/f1m.dat
    ```
  * run perf-record on the server side while stressing
    ```
    $ sudo perf record -g -t {{ a worker tid }} -- sleep 15
    ```
* Results
  * flamegraph
    * https://github.com/envoyproxy/envoy/pull/18485/commits/56152d2070ac0d505e888a36cae33101f31c048b
      * <img src="https://gist.githubusercontent.com/lkpdn/f0d875a9149154c0fe17ec56db86f743/raw/66952c3b8afda41ccd2a7dc05db653d524cde3c5/flamegraph_56152d2070ac.svg" width="400">
    * https://github.com/envoyproxy/envoy/pull/18485/commits/d4f01ac927faad35a9768e6d75fcaa00471408bf
      * <img src="https://gist.githubusercontent.com/lkpdn/f0d875a9149154c0fe17ec56db86f743/raw/66952c3b8afda41ccd2a7dc05db653d524cde3c5/flamegraph_d4f01ac927fa.svg" width="400">
  * zoom into the hot spot, seen in the former (https://github.com/envoyproxy/envoy/pull/18485/commits/56152d2070ac0d505e888a36cae33101f31c048b)
    ```                                                                                                                                                                                                                                                                                                                                                                                       
    Samples: 60K of event 'cycles', 4000 Hz, Event count (approx.): 30325457223
    Envoy::ThreadLocal::InstanceImpl::SlotImpl::get  /home/ubuntu/quic/envoy-dev/linux/arm64/build_envoy_debug/envoy [Percent: local period]
    Percent│
           │
           │
           │    Disassembly of section .text:
           │
           │    0000000003e1a040 <Envoy::ThreadLocal::InstanceImpl::SlotImpl::get()>:
           │    _ZN5Envoy11ThreadLocal12InstanceImpl8SlotImpl3getEv():
           │      stp  x29, x30, [sp, #-48]!
           │      str  x21, [sp, #16] 
           │      stp  x20, x19, [sp, #32]
           │      mov  x29, sp
           │      ldr  w20, [x0, #16]
           │      mov  x19, x8
           │      mov  x0, #0x10000                    // #65536
           │      movk x0, #0x3400
           │      nop  
           │      nop  
           │      mrs  x21, tpidr_el0
           │      ldrb w8, [x21, x0]
           │    ↓ cbz  w8, 80
           │34:   mov  x0, #0x10000                    // #65536
           │      movk x0, #0x33e0
           │      nop  
      0.00 │      nop  
           │      add  x8, x21, x0
           │      ldr  x8, [x8, #8] 
           │      add  x8, x8, x20, lsl #4
           │      ldp  x9, x8, [x8]
           │      stp  x9, x8, [x19]
           │    ↓ cbz  x8, 70
      0.00 │      add  x8, x8, #0x8
     28.44 │60:   ldxr x9, [x8]
     70.41 │      add  x9, x9, #0x1
           │      stxr w10, x9, [x8]
      0.05 │    ↑ cbnz w10, 60
      1.10 │70:   ldp  x20, x19, [sp, #32]
           │      ldr  x21, [sp, #16]
           │      ldp  x29, x30, [sp], #48
           │    ← ret  
           │80:   mov  x0, #0x10000                    // #65536
           │      movk x0, #0x3400
           │      nop
    ```




Risk Level: Low
Testing: 
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]
